### PR TITLE
[3.10] bpo-46197: Add isolated flag if currently isolated

### PR DESCRIPTION
**Pull request by kcdodd**
https://bugs.python.org/issue46197

Potential resolution of the issue introduced around Python 3.8.7 where ``ensurepip`` module would run the bootstrapped pip in a subprocess and no longer obeyed the ``isolated`` flag set when run by the ``venv`` module. This could cause pip to not install itself in the virtual environment if it found an installation in another path.

Approach taken is simple, to access the ``isolated`` flag of the current process, and, if it set, also add it to the sub-process. If the current process was not run as ``isolated`` the sub-process will not, restoring this particular case seen in Python <= 3.8.6.

This behavior was manually verified to work in the case that PYTHONPATH contains another installation of pip.

The test was attempted, but the ``test_venv`` test **appears to fail for another un-related issue** persisting from the last commit of the forked branch ``3.10`` (without the requested change). For clarity that the test does not indicate a problem with this request inspected traces of the failure is included below:

```
Traceback (most recent call last):
  File "<string>", line 6, in <module>
  File "/media/cdodd/box/projects/contrib/cpython/Lib/runpy.py", line 205, in run_module
    mod_name, mod_spec, code = _get_module_details(mod_name)
  File "/media/cdodd/box/projects/contrib/cpython/Lib/runpy.py", line 129, in _get_module_details
    spec = importlib.util.find_spec(mod_name)
  File "/media/cdodd/box/projects/contrib/cpython/Lib/importlib/util.py", line 103, in find_spec
    return _find_spec(fullname, parent_path)
  File "<frozen importlib._bootstrap>", line 945, in _find_spec
  File "/tmp/tmpr0dlj3q6/lib/python3.10/site-packages/_distutils_hack/__init__.py", line 79, in find_spec
    return method()
  File "/tmp/tmpr0dlj3q6/lib/python3.10/site-packages/_distutils_hack/__init__.py", line 100, in spec_for_pip
    if self.pip_imported_during_build():
  File "/tmp/tmpr0dlj3q6/lib/python3.10/site-packages/_distutils_hack/__init__.py", line 111, in pip_imported_during_build
    return any(
  File "/tmp/tmpr0dlj3q6/lib/python3.10/site-packages/_distutils_hack/__init__.py", line 112, in <genexpr>
    frame.f_globals['__file__'].endswith('setup.py')
KeyError: '__file__'
Traceback (most recent call last):
  File "/media/cdodd/box/projects/contrib/cpython/Lib/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/media/cdodd/box/projects/contrib/cpython/Lib/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/media/cdodd/box/projects/contrib/cpython/Lib/ensurepip/_uninstall.py", line 31, in <module>
    sys.exit(_main())
  File "/media/cdodd/box/projects/contrib/cpython/Lib/ensurepip/_uninstall.py", line 27, in _main
    return ensurepip._uninstall_helper(verbosity=args.verbosity)
  File "/media/cdodd/box/projects/contrib/cpython/Lib/ensurepip/__init__.py", line 231, in _uninstall_helper
    return _run_pip([*args, *reversed(_PACKAGE_NAMES)])
  File "/media/cdodd/box/projects/contrib/cpython/Lib/ensurepip/__init__.py", line 102, in _run_pip
    return subprocess.run( cmd, check = True ).returncode
  File "/media/cdodd/box/projects/contrib/cpython/Lib/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/tmp/tmpr0dlj3q6/bin/python', '-I', '-W', 'ignore::DeprecationWarning', '-c', '\nimport runpy\nimport sys\nsys.path = [] + sys.path\nsys.argv[1:] = [\'uninstall\', \'-y\', \'--disable-pip-version-check\', \'pip\', \'setuptools\']\nrunpy.run_module("pip", run_name="__main__", alter_sys=True)\n']' returned non-zero exit status 1.

ERROR

======================================================================
ERROR: test_with_pip (test.test_venv.EnsurePipTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/cdodd/box/projects/contrib/cpython/Lib/test/test_venv.py", line 551, in test_with_pip
    self.do_test_with_pip(True)
  File "/media/cdodd/box/projects/contrib/cpython/Lib/test/test_venv.py", line 517, in do_test_with_pip
    out, err = check_output([envpy,
  File "/media/cdodd/box/projects/contrib/cpython/Lib/test/test_venv.py", line 46, in check_output
    raise subprocess.CalledProcessError(
subprocess.CalledProcessError: Command '['/tmp/tmpr0dlj3q6/bin/python', '-W', 'ignore::DeprecationWarning', '-W', 'ignore::ImportWarning', '-I', '-m', 'ensurepip._uninstall']' returned non-zero exit status 1.
```

https://bugs.python.org/issue46197

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@kcdodd

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Lib/ensurepip/__init__.py [hunk 1]
```diff
@@ -90,8 +90,16 @@ def _run_pip(args, additional_paths=None):
 sys.argv[1:] = {args}
 runpy.run_module("pip", run_name="__main__", alter_sys=True)
 """
-    return subprocess.run([sys.executable, '-W', 'ignore::DeprecationWarning',
-                           "-c", code], check=True).returncode
+
+    cmd = [
+      sys.executable,
+      # run pip interpreter in isolated mode if current interpreter was isolated
```
Comment (merwok) [inline Lib/ensurepip/__init__.py]: probably mean python interpreter here

File: Lib/ensurepip/__init__.py [hunk 2]
```diff
@@ -90,8 +90,16 @@ def _run_pip(args, additional_paths=None):
 sys.argv[1:] = {args}
 runpy.run_module("pip", run_name="__main__", alter_sys=True)
 """
-    return subprocess.run([sys.executable, '-W', 'ignore::DeprecationWarning',
-                           "-c", code], check=True).returncode
+
+    cmd = [
+      sys.executable,
+      # run code in isolated mode if currently running isolated
+      *( [ '-I', ] if sys.flags.isolated else [] ),
```
Comment (merwok) [inline Lib/ensurepip/__init__.py]:
```suggestion
      *([ '-I'] if sys.flags.isolated else []),
```

**Comment by gvanrossum** [comment]
@kcdodd As stated in the issue, this should be rebased onto main and once accepted and merged there it can be backported. Probably a new PR is easiest.

**Comment by gvanrossum** [comment]
I recommend just clicking the CLA button. This is a side effect from the bpo->github migration. You'll only have to do it this once.

**Pull request closed without merging**